### PR TITLE
808 Checkpoint/Restart collection

### LINKED ISSUE #808: Restart collection from serialized files per element
**What Needs to be Done?**

- [ ] Add routines to checkpoint/restart collection elements.

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #808 

Has a test, working.

### CODE REVIEW

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/manager.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5499887&bid=18024896)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/manager.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5499887&bid=18024896)

File: src/vt/vrt/collection/manager.h [hunk 1]
```diff
@@ -875,6 +876,21 @@ struct CollectionManager
...
+   *
+   * \param[in] range the range of the collection to restart
+   * \param[in] file_base the base file name for the files to read
+   *
+   * \return the range of the collection
+   */
+  template <
+    typename ColT//, mapping::ActiveMapTypedFnType<typename ColT::IndexType> fn
+  >
+  CollectionProxyWrapType<ColT> restartFromFile(
```
Comment (PhilMiller) [inline src/vt/vrt/collection/manager.h:894]: We should name this `restoreFromFile` - there's no associated 'action' that would 'restart' anything

**Comment by PhilMiller** [comment]
This generally looks good, and pretty straightforward.

Something we may want to consider in `checkpointToFile` is putting the serialized size in the first `sizeof(size_t)` bytes, before the payload, so that we can compare that to the stat'd size of the file that we read in `restartFromFile`

**Comment by PhilMiller** [comment]
This does make an assumption of the array being dense in its index space. 

Looking ahead a little bit, will that be a problem if we wanted to use this in NimbleSM? Does that even care about preserving a collection instance across a hypothetical checkpoint?

**Comment by PhilMiller** [comment]
As we look toward large-scale runs, we're going to have a *lot* of objects. We may want to create subdirectories to hold limited numbers of objects each. I was going to suggest per-index-component, but that falls over if one of the index components is simply process rank, since even that gets too big. Instead, maybe we should hash the indices, and use a few bytes from the hash to set subdirectories?

**Comment by lifflander** [comment]
> This does make an assumption of the array being dense in its index space.
> 
> Looking ahead a little bit, will that be a problem if we wanted to use this in NimbleSM? Does that even care about preserving a collection instance across a hypothetical checkpoint?

I can fix that easily by just not failing if the file is not found during restart---assume it never existed.

**Comment by PhilMiller** [comment]
> > This does make an assumption of the array being dense in its index space.
> > Looking ahead a little bit, will that be a problem if we wanted to use this in NimbleSM? Does that even care about preserving a collection instance across a hypothetical checkpoint?
> 
> I can fix that easily by just not failing if the file is not found during restart---assume it never existed.

That seems a bit risky - How about each process recording the objects it knows it's the home process for, in an 'indices.$node'  file?

File: src/vt/vrt/collection/manager.impl.h [hunk 2]
```diff
@@ -3255,6 +3275,141 @@ IndexT CollectionManager::getRange(VirtualProxyType proxy) {
...
+) {
+  using IndexType = typename ColT::IndexType;
+  using DirectoryType = CollectionDirectory<IndexType>;
+
+  auto token = constructInsert<ColT>(range);
+
+  auto directory_name = makeMetaFilename<IndexType>(file_base, false);
+
+  if (access(directory_name.c_str(), F_OK) == -1) {
+    // file doesn't exist, true looking in sub-directory
```
Comment (PhilMiller) [inline src/vt/vrt/collection/manager.impl.h:3382]: Typo 'true' -> try

File: src/vt/vrt/collection/manager.impl.h [hunk 3]
```diff
@@ -3255,6 +3275,141 @@ IndexT CollectionManager::getRange(VirtualProxyType proxy) {
...
+CollectionManager::CollectionProxyWrapType<ColT>
+CollectionManager::restoreFromFile(
+  typename ColT::IndexType range, std::string const& file_base
+) {
+  using IndexType = typename ColT::IndexType;
+  using DirectoryType = CollectionDirectory<IndexType>;
+
+  auto token = constructInsert<ColT>(range);
+
+  auto directory_name = makeMetaFilename<IndexType>(file_base, false);
```
Comment (PhilMiller) [inline src/vt/vrt/collection/manager.impl.h:3379]: This variable is named contrary to its contents. `metadata_file_name` perhaps?

File: src/vt/vrt/collection/manager.impl.h (hunk 2 above)
Comment (lifflander) [reply src/vt/vrt/collection/manager.impl.h:3382]: Fixed

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_checkpoint.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5499887&bid=18024896)

File: src/vt/vrt/collection/manager.impl.h (hunk 3 above)
Comment (lifflander) [reply src/vt/vrt/collection/manager.impl.h]: Fixed

File: src/vt/vrt/collection/manager.h (hunk 1 above)
Comment (lifflander) [reply src/vt/vrt/collection/manager.h]: Done

**Comment by lifflander** [comment]
@PhilMiller I've addressed all your concerns and written a test for it!

**Comment by PhilMiller** [comment]
All lgtm

**Comment by lifflander** [comment]
@PhilMiller Good for beta.9?

**Comment by PhilMiller** [comment]
Yeah, good for the stable branch

**Pull request merged**